git_machete/git_operations.py

```diff
@@ -804,10 +804,53 @@ def is_equivalent_tree_reachable(
         )
 
         result = earlier_tree_hash in intermediate_tree_hashes
-        debug(f"result = {result}")
+        debug(f"intermediate tree hashes result = {result}")
+
+        if result:
+            self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = result
+            return result
+
+        # Let's try another way, a little more complex but takes into account the possibility
+        # that there were other commits between the fork point of the two branches and the squashed merge.
+        fork_point = self.get_merge_base(reachable_from_commit_hash, equivalent_to_commit_hash)
+        if not fork_point:
+            return False
+
+        equivalent_changeset = self._popen_git(
+            "diff",
+            fork_point,
+            equivalent_to_commit_hash
+        ).stdout
+        equivalent_patch_id = self.get_patch_id(equivalent_changeset)
+
+        if equivalent_patch_id is None and equivalent_changeset.strip() == '':
+            # Empty changeset means the branches are identical, so the tree is equivalent.
+            self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = True
+            return True
+
+        for logentry in self.get_commits_between(fork_point, reachable_from_commit_hash):
+            changeset = self._popen_git(
+                "diff",
+                logentry.hash + "~",
+                logentry.hash
+            ).stdout
+
+            patch_id = self.get_patch_id(changeset)
+            if patch_id == equivalent_patch_id:
+                result = True
+                break
+
+        debug(f"patch id result = {result}")
         self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = result
         return result
 
+    def get_patch_id(self, patch_contents: str) -> Optional[str]:
+        out = utils.get_non_empty_lines(self._popen_git("patch-id", input=patch_contents).stdout)
+
+        if len(out) == 0:
+            return None
+        return out[0].split(' ')[0]
```

Comment:
Pls mention why the split is necessary (the second part is commit hash, which we don't care about here)